File: DP/src/main/python/solution337.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rob(self, root: TreeNode) -> int:
        opt = {}
        opt[None] = [0, 0]

        def dfs(node):
            if node == None:
                return

            opt[node] = [0, 0]
            dfs(node.right)
            dfs(node.left)

            opt[node][0] = max(opt.get(node.right)[0], opt.get(node.right)[1]) + max(  # type: ignore
                opt.get(node.left)[0], opt.get(node.left)[1]  # type: ignore
            )
            opt[node][1] = opt.get(node.right)[0] + node.val + opt.get(node.left)[0]  # type: ignore

        dfs(root)

        # print(opt)
        return max(opt.get(root))

File: DP/src/main/java/Solution337.java
import java.util.HashMap;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution337 {

    private HashMap<TreeNode, int[]> opt;

    public int rob(TreeNode root) {
        opt = new HashMap<TreeNode, int[]>();
        opt.put(null, new int[] { 0, 0 });
        dfs(root);

        return Math.max(opt.get(root)[0], opt.get(root)[1]);
    }

    private void dfs(TreeNode node) {
        if (node == null)
            return;

        opt.put(node, new int[2]);

        dfs(node.left);
        dfs(node.right);

        opt.get(node)[0] = Math.max(
                opt.get(node.left)[0], opt.get(node.left)[1])
                + Math.max(
                        opt.get(node.right)[0], opt.get(node.right)[1]);
        opt.get(node)[1] = opt.get(node.left)[0] + node.val + opt.get(node.right)[0];

    }
}
